test/SgidClient.spec.ts

```diff
@@ -0,0 +1,332 @@
+        DEFAULT_RESPONSE_TYPE,
+      )
+      expect(actual.searchParams.get('redirect_uri')).toBe(MOCK_REDIRECT_URI)
+      // Important check is here
+      expect(actual.searchParams.get('nonce')).toBe(mockNonce)
+      expect(nonce).toBe(mockNonce)
+      expect(actual.searchParams.get('state')).toBe(mockState)
+      // Client ID, scope, response_type, redirect_uri, nonce, state
+      expect(actualNumSearchParams).toBe(6)
+    })
+
+    it('should generate authorisation URL correctly when nonce is null', () => {
+      const mockState = 'mockState'
+
+      const { url, nonce } = client.authorizationUrl(mockState, undefined, null)
+
+      const actual = new URL(url)
+      // Count number of search params
+      let actualNumSearchParams = 0
+      actual.searchParams.forEach(() => actualNumSearchParams++)
+      const expected = new URL(MOCK_AUTH_ENDPOINT)
+      expect(actual.host).toBe(expected.host)
+      expect(actual.pathname).toBe(expected.pathname)
+      expect(actual.searchParams.get('client_id')).toBe(MOCK_CLIENT_ID)
+      expect(actual.searchParams.get('scope')).toBe(DEFAULT_SCOPE)
+      expect(actual.searchParams.get('response_type')).toBe(
+        DEFAULT_RESPONSE_TYPE,
+      )
+      expect(actual.searchParams.get('redirect_uri')).toBe(MOCK_REDIRECT_URI)
+      // Important check is here
+      expect(actual.searchParams.get('nonce')).toBeNull()
+      expect(nonce).toBeUndefined()
+      expect(actual.searchParams.get('state')).toBe(mockState)
+      // Client ID, scope, response_type, redirect_uri, state
+      expect(actualNumSearchParams).toBe(5)
+    })
+
+    it('should generate authorisation URL correctly when redirectUri is provided', () => {
+      const mockState = 'mockState'
+      const mockRedirectUri = 'https://mockRedirectUri.com'
+
+      const { url, nonce } = client.authorizationUrl(
+        mockState,
+        undefined,
+        undefined,
+        mockRedirectUri,
+      )
+
+      const actual = new URL(url)
+      // Count number of search params
+      let actualNumSearchParams = 0
+      actual.searchParams.forEach(() => actualNumSearchParams++)
+      const expected = new URL(MOCK_AUTH_ENDPOINT)
+      expect(actual.host).toBe(expected.host)
+      expect(actual.pathname).toBe(expected.pathname)
+      expect(actual.searchParams.get('client_id')).toBe(MOCK_CLIENT_ID)
+      expect(actual.searchParams.get('scope')).toBe(DEFAULT_SCOPE)
+      expect(actual.searchParams.get('response_type')).toBe(
+        DEFAULT_RESPONSE_TYPE,
+      )
+      // Important check is here
+      expect(actual.searchParams.get('redirect_uri')).toBe(mockRedirectUri)
+      expect(actual.searchParams.get('nonce')).toBe(nonce)
+      expect(actual.searchParams.get('state')).toBe(mockState)
+      // Client ID, scope, response_type, redirect_uri, nonce, state
+      expect(actualNumSearchParams).toBe(6)
+    })
+
+    it('should throw when no redirectUri is provided', () => {
```

Comment:
current negative test case only covers no redirectUri is provided. Can consider state and scope is not provided respectively. Not super critical but for client facing library, we want to standardise all the error message.

My suggestion is to have all error messages as a constant file. This will help us in documentating this library when being published.

Reply:
so currently the library also doesn't throw if the state and scope are not provided. the server may throw, but imo testing that behaviour should be out of scope for testing this client library